### PR TITLE
contest timings

### LINKED ISSUE #103: Contest timings
There is no concept of contest timings right now.

- Keep record of when the contest is supposed to start and end (maybe store datetime in the database / question set).
- make sure that APIs always follow the timings. For example, if time is up then the check api should no longer work.

### PR DESCRIPTION
**Pull request by rishabhKalakoti**
- [x] contest timings system added
- [x] changed contest to defaultdict()
- [x] separate leaderboards for contests

fixes #103

### CODE REVIEW

File: server.py [hunk 1]
```diff
@@ -58,23 +66,24 @@ def contest(code,number):
 
 @app.get("/contest/<code>")
 def contest(code):
-    for contest in contests:
-        if(contest.code == code):
+    for contest_code, contest in contests.items():
```
Comment (theSage21) [inline server.py]: if you're using a dictionary, use a dictionary lookup instead of a loop

File: server.py [hunk 2]
```diff
@@ -58,23 +66,24 @@ def contest(code,number):
...
-def question(number):
-    statement = questions[number].statement
-    return template("index.html", question_number=number, question=statement)
-
+            start_time = contest.start_time
+            end_time = contest.end_time
+            if contest.start_time > datetime.datetime.now():
+                return "The contest had not started yet."
+    return template(
+        "contest.html", code=code, description=description, questions=questions,
```
Comment (theSage21) [inline server.py]: wouldn't it be easier to just pass in `contest=contest` and let the HTML use `contest.code`? Then while reading the html you know exactly what code it is that is being rendered there.

File: server.py [hunk 3]
```diff
@@ -11,26 +11,34 @@
 
 database_path = "submission_record.db"
 questions = {}
-contests = []
+contests = defaultdict()
```
Comment (theSage21) [inline server.py]: what is the default? I'd suggest a normal dictionary

File: server.py (hunk 2 above)
Comment (rishabhKalakoti) [reply server.py]: code is not part of contest as it is key, so i let `code` stay and passed `contest` for others

File: server.py [hunk 4]
```diff
@@ -43,31 +55,30 @@
...
-        if(contest.code == code):
-            description = contest.description
-            questions = contest.questions
-    return template("contest.html", code=code, description=description, questions=questions)
-
-@app.get("/question/<number>")
-def question(number):
-    statement = questions[number].statement
-    return template("index.html", question_number=number, question=statement)
+    if not code in contests.keys():
```
Comment (theSage21) [inline server.py]:
```suggestion
    if not code in contests:
```

if you use `.keys()` it makes a list-ish item and then looks up in that. Instead use the constant time lookup property of dictionaries

**Comment by rishabhKalakoti** [comment]
tested :)

File: server.py [hunk 5]
```diff
@@ -91,6 +93,36 @@ def server_static(filepath):
...
+                                and attempt.time >= contests[code].start_time
+                            )
+                        ]
+                    )
+                ),
+            )
+            for user, submissions in submission_record.items()
+        ]
+    order.sort(key=lambda x: x[1], reverse=True)
+    order = [entry for entry in order if entry[1] > 0]
```
Comment (rishabhKalakoti) [inline server.py]: @theSage21 can you suggest some way for me to remove line 121?
`order = [entry for entry in order if entry[1] > 0]`

Comment (theSage21) [reply server.py]: umm. you could combine both of them.
```python
order = [entry for entry in sorted(order, key=lambda x: x[1], reverse=True) if entry[1] > 0]
```

Comment (theSage21) [reply server.py]: another way is that when you are generating the order list don't append things that are  <= 0.
```python
order = [entry for entry in (
                                          (user, <some complex calculation)
                                          for something in something
                                         ) if entry[1] > 0]
order.sort(<fancy arguments>)
```
the inner iteration being `()` makes it a generator which means nothing is stored. Then in the outer loop, once the condition is met it is added to the list. So actually you don't add the entry if the condition is met. then you just sort it.

The current method is ok though. It's easy to read and works

Comment (rishabhKalakoti) [reply server.py]: The contents of `order` look complicated already, so... I will go with the current one for now.

Comment (rishabhKalakoti) [reply server.py]: Let me know if there is something else I can add to this PR

File: server.py [hunk 6]
```diff
@@ -43,31 +55,32 @@
...
 
+
 @app.get("/dashboard")
 def dashboard():
     return template("dashboard.html", contests=contests)
 
+
 @app.get("/contest/<code>/<number>")
-def contest(code,number):
+def contest(code, number):
```
Comment (theSage21) [inline server.py]: we should not be able to see contest if it is not open

File: server.py [hunk 7]
```diff
@@ -43,31 +55,32 @@
...
-            questions = contest.questions
-    return template("contest.html", code=code, description=description, questions=questions)
-
-@app.get("/question/<number>")
-def question(number):
-    statement = questions[number].statement
-    return template("index.html", question_number=number, question=statement)
+    if not code in contests:
+        return "Contest does not exist"
+    if contests[code].start_time > datetime.datetime.now():
```
Comment (theSage21) [inline server.py]: we shouldn't be able to attempt after a contest is closed

Comment (rishabhKalakoti) [reply server.py]: Most online Judges keep contest questions open to attempt after the contest.
The score for submissions after the contest ends is not considered in the ranking.
I kept it the same way. We could display a message that contest has ended. Or do you want me to close the contest after its time is over.

File: server.py (hunk 6 above)
Comment (rishabhKalakoti) [reply server.py]:
```
if contests[code].start_time > datetime.datetime.now():
        return "The contest had not started yet."
```
These lines in the code already do that.

Comment (rishabhKalakoti) [reply server.py]: I have not added a handler in for the questions pattern yet though.

File: server.py (hunk 7 above)
Comment (theSage21) [reply server.py]: sure sure. keep it open but don't record the score then? The code submission api needs a change I think

File: server.py (hunk 6 above)
Comment (theSage21) [reply server.py]: umm. that's in another api no?

Comment (rishabhKalakoti) [reply server.py]: oh, sorry didnt check the position, will do it right away. This is in the questions handler

File: server.py (hunk 7 above)
Comment (rishabhKalakoti) [reply server.py]: We are recording just the submissions to problems. The score is calculated while displaying the ranking and it calculates the score only for the timings the contest is on

File: server.py (hunk 6 above)
Comment (rishabhKalakoti) [reply server.py]: Done.

**Comment by theSage21** [comment]
All right then.

Sorry I'm not able to devote full attention to this. Bogged with reading
backlog

File: server.py (hunk 7 above)
Comment (rishabhKalakoti) [reply server.py]: This one's fine then, I guess

**Comment by rishabhKalakoti** [comment]
@theSage21 , So... is this one good to go yet?

**Pull request merged**